src/common/ast-path.js

```diff
@@ -1,35 +1,79 @@
+import assert from "node:assert";
 import getLast from "../utils/get-last.js";
 
-function getNodeHelper(path, count) {
-  const stackIndex = getNodeStackIndexHelper(path.stack, count);
-  return stackIndex === -1 ? null : path.stack[stackIndex];
-}
+class AstPath {
+  constructor(value) {
+    this.stack = [value];
+  }
 
-function getNodeStackIndexHelper(stack, count) {
-  for (let i = stack.length - 1; i >= 0; i -= 2) {
-    const value = stack[i];
-    if (value && !Array.isArray(value) && --count < 0) {
-      return i;
+  get key() {
+    const { stack } = this;
+    const { length } = stack;
+    let key = stack[length - 2];
+    if (typeof key === "number") {
+      assert(Array.isArray(stack[length - 3]));
+      key = stack[length - 4];
     }
+    return key ?? null;
   }
-  return -1;
-}
 
-class AstPath {
-  constructor(value) {
-    this.stack = [value];
+  get index() {
+    const { stack } = this;
+    const { length } = stack;
+    const index = stack[length - 2];
+    return typeof index === "number" ? index : null;
+  }
+
+  get node() {
+    return getLast(this.stack);
+  }
+
+  get parent() {
+    return this.getNode(1);
+  }
+
+  get grandparent() {
+    return this.getNode(2);
+  }
+
+  get next() {
+    return this.siblings[this.index + 1];
+  }
+
+  get previous() {
+    return this.siblings[this.index - 1];
+  }
+
+  get siblings() {
+    const { stack } = this;
+    const { length } = stack;
+    const array = stack[length - 3];
+    assert(Array.isArray(array));
+    return array;
+  }
+
+  get isInArray() {
+    return this.index !== null;
```

Comment:
I was hoping `Array.isArray[this.stack.length - 3]`, and other getters call this one first. More reliable.